docs/man/man4/ttyld.4
.TH TTYLD 4
.CT 2 comm_term
.SH NAME
tty_ld \- terminal processing line discipline
.SH SYNOPSIS
.B #include <sys/ttyio.h>
.SH DESCRIPTION
.I Tty_ld
is usually inserted
into a stream
connected to a terminal device.
It gathers input into lines,
handles special characters like erase, kill, and interrupt,
inserts output delays,
and the like.
It does not deal with hardware parameters
such as speed and parity;
see
.IR tty (4)
for such matters.
.PP
Certain special characters have particular meaning on input.
These characters are not passed to a program
except in raw mode, where they lose their special character.
It is possible to change these characters from the default.
.PP
The
.I erase
character
(backspace by default)
erases the last-typed character.
It will not erase beyond the beginning of a line or
an end-of-file character.
.PP
The
.I kill
character
(default
.BR @ )
erases the entire preceding part of the line, but not beyond
an end-of-file character.
.PP
The
.I end-of-file
character
(default
.RB control- d )
causes any characters
waiting to be read
to be passed immediately
to the program,
without waiting for newline.
The
end-of-file character itself is discarded.
Thus if the end-of-file character
occurs at the beginning of a line,
there are no characters waiting,
and zero characters will be passed back;
this is
the standard end-of-file indication.
.PP
The
.I escape
character
.RB ( \e )
escapes a following erase, kill, or end-of-file character
and allows it to be treated as ordinary data.
.PP
The
.I interrupt
character
(default
DEL)
is not passed to a program but sends
signal
.L SIGINT
to any processes in the process group of the stream;
see
.IR signal (2)
and
.IR stream (4).
.PP
The
.I quit
character
(default
FS,
.RB control- \e )
sends signal
.BR SIGQUIT .
.PP
The
.I stop
character
(default DC3,
.RB control- s )
delays printing on the terminal
until something is typed in.
.PP
The
.I start
character
(default DC1,
.RB control- q )
restarts printing after 
a stop character without generating
any input to a program.
.PP
Two
.IR ioctl (2)
calls affect these characters:
.TF TIOCGETC
.TP
.B TIOCGETC
The argument points to a
.B tchars
structure to be filled in with current settings.
.TP
.B TIOCSETC
The argument points to a
.B tchars
structure from which the characters are set.
.PD
.PP
The
.B tchars
structure, as defined in
.LR <sys/ttyio.h> ,
is
.PP
.EX
.ta \w'struct 'u +\w'tchars 'u +\w't_startc; 'u
struct tchars {
	char	t_intrc;	/* interrupt */
	char	t_quitc;	/* quit */
	char	t_startc;	/* start output */
	char	t_stopc;	/* stop output */
	char	t_eofc;	/* end-of-file */
	char	t_brkc;	/* input delimiter (like nl) */
};
.EE
.PP
A character value of 0377
eliminates the effect of that character.
The
.L t_brkc
character, by default 0377,
acts like a new-line in that it terminates a line,
is echoed, and is passed to the program.
The stop and start characters may be the same,
to produce a toggle effect.
It is counterproductive to make
other special characters (including erase and kill)
identical.
.PP
Two
.I ioctl
calls affect other terminal processing parameters:
.TF TIOCGETP
.TP
.B TIOCGETP
The argument points to a
.B sgttyb
structure to be filled in with the current settings.
.TP
.B TIOCSETP
The argument points to a 
.B sgttyb
structure from which the parameters are set.
.PD
.PP
The
.B sgttyb
structure, as
defined in
.LR <sys/ttyio.h> ,
is
.EX
.ta \w'struct 'u +\w'sgttyb 'u +\w'sg_ispeed 'u
struct	sgttyb	{
	char	sg_ispeed;	/* unused */
	char	sg_ospeed;	/* unused */
	char	sg_erase;	/* erase character */
	char	sg_kill;	/* kill character */
	short	sg_flags;	/* mode flags */
};
.EE
.PP
The flag bits are
.PP
.ta \w'\f5ALLDELAY\fP 'u +\w'\f50100000\fP 'u
.nf
\f5ALLDELAY	0177400\fP	Delay algorithm selection
\f5VTDELAY	0040000\fP	Form-feed and vertical-tab delays:
\f5FF0	0\fP
\f5FF1	0040000\fP
\f5CRDELAY	0030000\fP	Carriage-return delays:
\f5CR0	0\fP
\f5CR1	0010000\fP
\f5CR2	0020000\fP
\f5CR3	0030000\fP
\f5TBDELAY	0006000\fP	Tab delays:
\f5TAB0	0\fP
\f5TAB1	0002000\fP
\f5TAB2	0004000\fP
\f5XTABS	0006000\fP
\f5NLDELAY	0001400\fP	New-line delays:
\f5NL0	0\fP
\f5NL1	0000400\fP
\f5NL2	0001000\fP
\f5NL3	0001400\fP
\f5RAW	0000040\fP	Raw mode: wake up on all characters
\f5CRMOD	0000020\fP	Map CR into LF; echo LF or CR as CR-LF
\f5ECHO	0000010\fP	Echo (full duplex)
\f5LCASE	0000004\fP	Map upper case to lower on input
\f5CBREAK	0000002\fP	Return each character as soon as typed
\f5TANDEM	0000001\fP	Automatic flow control
.DT
.fi
.PP
The delay bits specify how long
transmission stops to allow for mechanical or other movement
when certain characters are sent to the terminal.
In all cases a value of 0 indicates no delay.
.PP
If a form-feed/vertical tab delay is specified,
it lasts for about 2 seconds.
.PP
Carriage-return delay type 1 lasts about .08 seconds;
type 2 about .16 seconds;
type 3 about .32 seconds.
.PP
New-line delay type 1 is supposed to be for the Teletype model 37;
type 2 is about .10 seconds.
.PP
Tab delay type 1 is supposed to be for the Teletype model 37.
Type 3, called 
.LR XTABS ,
is not a delay at all but causes tabs to be replaced
by the appropriate number of spaces on output.
.PP
In
.B RAW
mode,
every character is passed immediately
to the program without waiting until a full line has been typed.
No erase or kill processing is done;
the end-of-file, interrupt, and quit characters
are not treated specially.
There are no delays and no echoing, and no replacement of
one character for another.
.PP
.L CRMOD
causes input carriage returns to be turned into
new-lines;
input of either CR or LF causes CR-LF both to
be echoed
(for terminals without a new-line function).
.PP
.L CBREAK
is a sort of half-cooked mode.
Programs read each character as soon as typed, instead
of waiting for a full line,
but quit and interrupt work, and output delays
.LR CRMOD ,
.LR XTABS ,
and
.LR ECHO
work normally.
On the other hand there is no erase or kill,
and no special treatment of
.L \e
or end-of-file.
.PP
.L TANDEM
mode causes the system to transmit
the stop character whenever the input
queue is in danger of overflowing,
and the start character
when the input queue has drained sufficiently.
It is useful for flow control when the `terminal'
is actually another machine that obeys the conventions.
.SH SEE ALSO
.IR getty (8), 
.IR stty (1), 
.IR signal (2), 
.IR ioctl (2), 
.IR stream (4),
.IR tty (4)
.SH BUGS
The
escape character
cannot be changed.


v10/sys/io/ttyld.c
#include "sys/param.h"
#include "sys/stream.h"
#include "sys/ttyio.h"
#include "sys/ttyld.h"
#include "sys/conf.h"

extern	char	partab[];

#define	CANBSIZ	256		/* size of largest input line */

extern	struct	ttyld	ttyld[];
extern	int ttycnt;

char	maptab[] = {
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,'|',000,000,000,000,000,'`',
	'{','}',000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,000,000,000,000,
	000,000,000,000,'\\',000,'~',000,
	000,'A','B','C','D','E','F','G',
	'H','I','J','K','L','M','N','O',
	'P','Q','R','S','T','U','V','W',
	'X','Y','Z',000,000,000,000,000,
};

struct block	*canonblock();

long	ttyopen();
int	ttyclose(), ttyldin(), ttyinsrv(), ttyosrv();
static struct qinit ttrinit = { ttyldin, ttyinsrv, ttyopen, ttyclose, 3*1024, 60};
static struct qinit ttwinit = { putq, ttyosrv, ttyopen, ttyclose, 3*1024, 200};
struct streamtab ttystream = { &ttrinit, &ttwinit};

/*
 * TTY open
 */
long
ttyopen(qp, dev)
register struct queue *qp;
{
	register struct ttyld *tp;
	static struct tchars tchars = {CINTR,CQUIT,CSTART,CSTOP,CEOT,0377};

	if (qp->ptr)				/* already attached */
		return(1);
	for (tp = ttyld; tp->t_state&TTUSE; tp++)
		if (tp >= &ttyld[ttycnt-1])
			return(0);
	tp->t_state = TTUSE;
	tp->t_flags = ECHO|CRMOD;
	tp->t_delct = 0;
	tp->t_col = 0;
	tp->t_erase = CERASE;
	tp->t_kill = CKILL;
	tp->t_chr = tchars;
	qp->ptr = (caddr_t)tp;
	qp->flag |= QDELIM|QNOENB;
	WR(qp)->ptr = (caddr_t)tp;
	return(1);
}

ttyclose(qp)
struct queue *qp;
{
	struct ttyld *tp = (struct ttyld *)qp->ptr;

	if (tp->t_state & TTSTOP)
		putctl(WR(qp)->next, M_START);	/* what else can we do? */
	tp->t_state = 0;
}

/*
 * Queue put procedure for tty input
 */
ttyldin(q, bp)
struct queue *q;
register struct block *bp;
{
	register struct ttyld *tp;
	register c;
	register struct queue *wrq = WR(q);	/* writer side */
	int escape, flags;

	tp = (struct ttyld *)q->ptr;
	flags = tp->t_flags;
	bp->class &= ~S_DELIM;
	if (bp->type!=M_DATA) {
		switch(bp->type) {

		case M_BREAK:
			if (tp->t_flags&RAW) {		/* speed-change hack*/
				bp->type = M_DATA;
				bp->class &= ~S_DELIM;
				if (bp->wptr<bp->lim)
					*bp->wptr++ = '\0';
				break;
			}
			ttysig(q, SIGINT);
			freeb(bp);
			return;

		case M_HANGUP:
		case M_IOCACK:
		case M_IOCNAK:
			(*q->next->qinfo->putp)(q->next, bp);
			return;

		case M_IOCTL:
			ttldioc(WR(q), bp, q, 1);
			return;
		}
		flags |= RAW;
	}
	if (tp->t_flags&TANDEM && tp->t_state&TTBLOCK
	 && q->count <= q->qinfo->lolimit) {
		tp->t_state &= ~TTBLOCK;
		putd(putq, WR(q), tp->t_chr.t_startc);
	}
	if (flags&RAW) {
		if ((q->next->flag&QFULL)==0 && q->count==0)
			(*q->next->qinfo->putp)(q->next, bp);
		else
			putq(q, bp);
		return;
	}
	while (bp->rptr<bp->wptr) {
		c = *bp->rptr++ & 0177;
		if (tp->t_state&TTSTOP) {
			if (c!=tp->t_chr.t_stopc
			 || tp->t_chr.t_stopc==tp->t_chr.t_startc) {
				tp->t_state &= ~TTSTOP;
				putctl(wrq->next, M_START);
			}
		} else {
			if (c==tp->t_chr.t_stopc) {
				tp->t_state |= TTSTOP;
				putctl(wrq->next, M_STOP);
			}
		}
		if (c==tp->t_chr.t_stopc || c==tp->t_chr.t_startc)
			continue;
		if (c==tp->t_chr.t_intrc) {
			ttysig(q, SIGINT);
			continue;
		}
		if (c==tp->t_chr.t_quitc) {
			ttysig(q, SIGQUIT);
			continue;
		}
		if (c=='\r' && tp->t_flags&CRMOD)
			c = '\n';
		if (tp->t_flags&LCASE && c>='A' && c<='Z')
			c += 'a'-'A';
		escape = 0;
		if (tp->t_flags & CBREAK) {
			if ((q->next->flag&QFULL)==0 && q->count==0)
				putd(q->next->qinfo->putp, q->next, c);
			else
				putd(putq, q, c);
		} else {
			if (tp->t_state&TTESC) {
				escape = 1;
				c |= 0200;
			}
			if (c == '\\')
				tp->t_state |= TTESC;
			else {
				tp->t_state &= ~TTESC;
				if (c == ('\\'|0200)) {
					c &= 0177;
					tp->t_state |= TTESC;
				}
				/* ttyhog? */
				if ((q->flag&QFULL) == 0 || (c=='\n' && tp->t_delct==0))
					putd(putq, q, c);
				else
					c = '\007';
			}
			if ((c&0177)=='\n'||c==tp->t_chr.t_eofc
			 ||c==tp->t_chr.t_brkc) {
				register struct block *bp1;
				if (bp1 = allocb(1)) {
					bp1->class |= S_DELIM;
					tp->t_delct++;
					putq(q, bp1);
				}
				qenable(q);
			}
		}
		if (tp->t_flags&TANDEM && (tp->t_state&TTBLOCK) == 0
		 && q->count >= (q->qinfo->limit+q->qinfo->lolimit)/2 ) {
			q->next->flag |= QWANTW;
			tp->t_state |= TTBLOCK;
			putctl1d(wrq, M_DATA, tp->t_chr.t_stopc);
		}
		if (tp->t_flags&ECHO && (wrq->flag&QFULL)==0) {
			c &= 0177;
			if(bp->rptr == bp->wptr)
				putctl1d(wrq, M_DATA, c);
			else
				putd(wrq->qinfo->putp, wrq, c);
			if (c==tp->t_kill && (tp->t_flags&CBREAK)==0
			 && !escape)
				putctl1d(wrq, M_DATA, '\n');
		}
	}
	freeb(bp);
}

/*
 * tty input server processing.  Erase-kill and escape processing;
 * gathering into lines.
 */

ttyinsrv(q)
register struct queue *q;
{
	register struct ttyld *tp;
	register struct block *bp, *bp1;

	tp = (struct ttyld *)q->ptr;
	if (q->next->flag&QFULL)
		return;
	if (tp->t_flags&(CBREAK|RAW)) {
		while ((q->next->flag&QFULL)==0 && (bp = getq(q)))
			(*q->next->qinfo->putp)(q->next, bp);
	} else {
		while (tp->t_delct && q->first) {
			bp1 = allocb(CANBSIZ);
			if (bp1==NULL)
				return;
			while (bp = getq(q)) {
				bp1 = canonblock(q, bp, bp1, tp);
				if (bp1->class&S_DELIM) {
					tp->t_delct--;
					break;
				}
			}
			(*q->next->qinfo->putp)(q->next, bp1);
		}
	}
	if (tp->t_flags&TANDEM && tp->t_state&TTBLOCK
	 && q->count <= q->qinfo->lolimit) {
		tp->t_state &= ~TTBLOCK;
		putd(putq, WR(q), tp->t_chr.t_startc);
	}
}

/*
 * canonicalize bp into bp1, noticing delimiter.  bp is freed.
 */
struct block *
canonblock(q, bp, bp1, tp)
register struct queue *q;
register struct block *bp, *bp1;
register struct ttyld *tp;
{
	register c;

	while (bp->rptr<bp->wptr) {
		if (bp1->wptr >= bp1->lim-1) {
			(*q->next->qinfo->putp)(q->next, bp1);
			bp1 = allocb(CANBSIZ);
		}
		c = *bp->rptr++;
		if ((c&0200) == 0) {	/* not escaped */
			if (c == tp->t_erase) {
				if (bp1->wptr > bp1->rptr)
					bp1->wptr--;
				continue;
			}
			if (c == tp->t_kill) {
				bp1->wptr = bp1->rptr;
				continue;
			}
			if (c == tp->t_chr.t_eofc)
				continue;
		} else {
			c &= 0177;
			if (tp->t_flags&LCASE && maptab[c])
				c = maptab[c];
			else if (c==tp->t_erase || c==tp->t_kill
			      || c==tp->t_chr.t_eofc)
				;
			else
				*bp1->wptr++ = '\\';
		}
		*bp1->wptr++ = c;
	}
	if (bp->class&S_DELIM)
		bp1->class |= S_DELIM;
	freeb(bp);
	return(bp1);
}

/*
 * TTY write processing: delays, tabs, CR/NL and the like.
 */
ttyosrv(q)
register struct queue *q;
{
	register struct ttyld *tp;
	register struct block *bp;

	tp = (struct ttyld *)q->ptr;
	while (bp = getq(q)) {
		switch(bp->type) {

		default:
			freeb(bp);
			continue;

		case M_IOCTL:
			if (q->next->flag & QFULL) {
				putbq(q, bp);
				return;
			}
			ttldioc(q, bp, RD(q), 0);
			continue;

		case M_FLUSH:
			flushq(q, 0);
		case M_IOCNAK:		/* flow through */
		case M_IOCACK:
			(*q->next->qinfo->putp)(q->next, bp);
			continue;

		case M_DATA:
		case M_BREAK:
			if (q->next->flag & QFULL) {
				putbq(q, bp);
				return;
			}
			if (tp->t_flags&RAW || bp->type==M_BREAK) {
				(*q->next->qinfo->putp)(q->next, bp);
			} else
				outconv(q, bp);
			continue;
		}
	}
}

outconv(q, ibp)
struct queue *q;
register struct block *ibp;
{
	register struct ttyld *tp;
	register struct block *obp = NULL;
	register c;
	register count, ctype;

	tp = (struct ttyld *)q->ptr;
   more:
	while (ibp->rptr < ibp->wptr) {
		if (obp==NULL || obp->wptr >= obp->lim) {
			if (obp)
				(*q->next->qinfo->putp)(q->next, obp);
			if (q->next->flag&QFULL || (obp=allocb(QBSIZE))==NULL) {
				putbq(q, ibp);
				return;
			}
		}
		/*
		 * The following dance is an inner loop
		 */
		count = ibp->wptr - ibp->rptr;
		if ((c = obp->lim - obp->wptr) < count)
			count = c;
		while ((ctype = partab[c = *ibp->rptr++ & 0177] & 077) == 0) {
			tp->t_col++;
			*obp->wptr++ = c;
			if (--count <= 0)
				goto more;
		}
		if (c=='\t' && (tp->t_flags&TBDELAY)==XTABS) {
			for (;;) {
				*obp->wptr++ = ' ';
				tp->t_col++;
				if ((tp->t_col & 07) == 0)	/* every 8 */
					break;
				if (obp->wptr >= obp->lim) {
					ibp->rptr--;
					break;
				}
			}
			continue;
		}

		/*
		 * turn <nl> to <cr><lf> if desired.
		 */
		if (c=='\n' && tp->t_flags&CRMOD) {
			if ((tp->t_state&TTCR)==0) {
				tp->t_state |= TTCR;
				c = '\r';
				ctype = partab['\r'] & 077;
				--ibp->rptr;
			} else
				tp->t_state &= ~TTCR;
		}
		/*
		 * store character
		 */
		*obp->wptr++ = c;
		/*
		 * Calculate delays and column movement
		 */
		count = 0;
		switch (ctype) {

		/* ordinary */
		case 0:
			tp->t_col++;
			break;
	
		/* non-printing */
		case 1:
			break;
	
		/* backspace */
		case 2:
			if (tp->t_col)
				tp->t_col--;
			break;
	
		/* newline */
		case 3:
			ctype = (tp->t_flags >> 8) & 03;
			if(ctype == 1) { /* tty 37 */
				if (tp->t_col)
					count = max(((unsigned)tp->t_col>>4) + 3, (unsigned)6);
			} else if (ctype == 2)	/* vt05 */
				count = 6;
			if ((tp->t_flags&CRMOD)==0)
				tp->t_col = 0;
			break;
	
		/* tab */
		case 4:
			ctype = (tp->t_flags >> 10) & 03;
			if(ctype == 1) { /* tty 37 */
				count = 1 - (tp->t_col | ~07);
				if (count < 5)
					count = 0;
			}
			tp->t_col |= 07;
			tp->t_col++;
			break;
	
		/* vertical motion */
		case 5:
			if(tp->t_flags & VTDELAY)
				count = 127;
			break;
	
		/* carriage return */
		case 6:
			ctype = (tp->t_flags >> 12) & 03;
			if (ctype == 1) 	 /* tn 300 */
				count = 5;
			else if (ctype == 2)	/* ti 700 */
				count = 10;
			else if (ctype == 3)
				count = 20;
			tp->t_col = 0;
			break;
		}
		if (count) {
			(*q->next->qinfo->putp)(q->next, obp);
			putctl1(q->next, M_DELAY, count);
			obp = NULL;
		}
	}
	if (obp) {
		obp->class |= ibp->class&S_DELIM;
		(*q->next->qinfo->putp)(q->next, obp);
	} else if (ibp->class&S_DELIM)
		putctld(q->next, M_DATA);
	freeb(ibp);
}

/*
 * Reader generates a signal and passes it up
 */
ttysig(q, sig)
register struct queue *q;
{
	register struct ttyld *tp = (struct ttyld *)q->ptr;

	flushq(q, 0);			/* flush reader */
	flushq(WR(q), 0);
	tp->t_state &= ~TTESC;
	tp->t_delct = 0;
	putctl(q->next, M_FLUSH);
	putctl1(q->next, M_SIGNAL, sig);
	putctl(WR(q)->next, M_FLUSH);
}

ttldioc(q, bp, rdq, fromdev)
register struct block *bp;
struct queue *q, *rdq;
{
	register struct ttyld *tp;
	register struct sgttyb *sp;
	int s;

	sp = (struct sgttyb *)stiodata(bp);
	tp = (struct ttyld *)q->ptr;
	switch (stiocom(bp)) {

	/*
	 * Set new parameters
	 */
	case TIOCSETP:
	case TIOCSETN:
		s = spl6();
		if (sp->sg_flags & (RAW|CBREAK)
		  && (rdq->next->flag&QFULL)==0) {
			register struct block *bp1;
			ttyinsrv(rdq);
			while (bp1 = getq(rdq))
				(*rdq->next->qinfo->putp)(rdq->next, bp1);
		}
		tp->t_erase = sp->sg_erase;
		tp->t_kill = sp->sg_kill;
		tp->t_flags = sp->sg_flags;
		splx(s);
		bp->type = M_IOCACK;
		if (tp->t_flags & (RAW|CBREAK))
			rdq->flag &= ~(QDELIM|QNOENB);
		else
			rdq->flag |= QDELIM|QNOENB;
		if (tp->t_flags & RAW && tp->t_state & TTSTOP) {
			putctl(q->next, M_START);	/* else it is stuck forever */
			tp->t_state &=~ TTSTOP;
		}
		break;

	/*
	 * Send current parameters to user
	 */
	case TIOCGETP:
		sp->sg_erase = tp->t_erase;
		sp->sg_kill = tp->t_kill;
		sp->sg_flags = tp->t_flags;
		sp->sg_ispeed = sp->sg_ospeed = B9600;
		bp->wptr = bp->rptr+sizeof(struct sgttyb)+STIOCHDR;
		bp->type = M_IOCACK;
		break;

	/*
	 * Set and fetch special characters
	 */
	case TIOCSETC:
		tp->t_chr = *(struct tchars *)((struct stioctl *)bp->rptr)->data;
		bp->wptr = bp->rptr;
		bp->type = M_IOCACK;
		break;

	case TIOCGETC:
		*(struct tchars *)((struct stioctl *)bp->rptr)->data = tp->t_chr;
		bp->wptr = bp->rptr+sizeof(struct tchars)+STIOCHDR;
		bp->type = M_IOCACK;
		break;

	default:
		if (fromdev) {
			bp->type = M_IOCACK;
			qreply(rdq, bp); /* reply to device side */
		} else
			(*q->next->qinfo->putp)(q->next, bp); /* pass to device */
		return;

	}
	if (fromdev)
		qreply(rdq, bp);	/* to device side */
	else
		qreply(q, bp);		/* to process side */
}
